File: src/scala/InheritancePolymorphism.scala
package typeclass

trait Base:
    def consume(): Unit
    def communicate(): Unit

class Dog extends Base:
    def consume() = println("eating dog food")
    def communicate() = println("woof")

class Person extends Base:
    def consume() = println("eating pizza")
    def communicate() = println("hello")

class Robot extends Base:
    def consume() = println("recharging")
    def communicate() = println("operational")

class Slug extends Base:
    def consume() = println("eating grass")
    def communicate() = {}

def demo(instance: Base): Unit =
    instance.consume()
    instance.communicate()

@main def main() =
    List(Dog(), Person(), Robot(), Slug()).foreach(demo(_))

File: src/kotlin/src/main/kotlin/InheritancePolymorphism.kt
package inheritancepolymorphism

abstract class Base {
    abstract fun eat()
    abstract fun speak()
}

class Dog: Base() {
    override fun eat() { println("eating dog food") }
    override fun speak() { println("woof") }
}

class Person: Base()  {
    override fun eat() { println("eating pizza") }
    override fun speak() { println("hello") }
}

class Robot: Base()  {
    override fun eat() { println("charging") }
    override fun speak() { println("operational") }
}

class Slug: Base()  {
    override fun eat() { println("eating grass") }
    override fun speak() { }
}

fun nourish(subject: Base) {
    subject.eat()
    subject.speak()
}

fun main() {
    listOf(
        Dog(), Person(), Robot(), Slug()
    ).forEach { nourish(it) }
}
